### PR TITLE
[SPARK-11078] Ensure spilling tests actually spill

### PR DESCRIPTION
**Pull request by andrewor14**
#9084 uncovered that many tests that test spilling don't actually spill. This is a follow-up patch to fix that to ensure our unit tests actually catch potential bugs in spilling. The size of this patch is inflated by the refactoring of `ExternalSorterSuite`, which had a lot of duplicate code and logic.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala [hunk 1]
```diff
@@ -18,535 +18,91 @@
...
+    val size = 1000
+    val conf = createSparkConf(loadDefaults = true, kryo = false)
+    conf.set("spark.shuffle.spill.numElementsForceSpillThreshold", (size / 2).toString)
     sc = new SparkContext("local-cluster[1,1,1024]", "test", conf)
 
     def createCombiner(i: String): ArrayBuffer[String] = ArrayBuffer[String](i)
     def mergeValue(buffer: ArrayBuffer[String], i: String): ArrayBuffer[String] = buffer += i
     def mergeCombiners(buffer1: ArrayBuffer[String], buffer2: ArrayBuffer[String])
-      : ArrayBuffer[String] = buffer1 ++= buffer2
+    : ArrayBuffer[String] = buffer1 ++= buffer2
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala]: nit:
```
def mergeCombiners(
    buffer1: ArrayBuffer[String],
    buffer2: ArrayBuffer[String]): ArrayBuffer[String]
```

**Comment by cloud-fan** [comment]
LGTM over all

File: core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala]: oops I screwed that up

**Comment by andrewor14** [comment]
OK, merging.

**Comment by srowen** [comment]
@andrewor14 looks like we still have some failures of the form...

```
- spilling with compression *** FAILED ***
  java.lang.Exception: Test failed with compression using codec org.apache.spark.io.LZ4CompressionCodec:

assertion failed: expected groupByKey to spill, but did not
  at scala.Predef$.assert(Predef.scala:170)
  at org.apache.spark.TestUtils$.assertSpilled(TestUtils.scala:170)
  at org.apache.spark.util.collection.ExternalAppendOnlyMapSuite.org$apache$spark$util$collection$ExternalAppendOnlyMapSuite$$testSimpleSpilling(ExternalAppendOnlyMapSuite.scala:253)
  at org.apache.spark.util.collection.ExternalAppendOnlyMapSuite$$anonfun$10$$anonfun$apply$mcV$sp$8.apply(ExternalAppendOnlyMapSuite.scala:218)
  at org.apache.spark.util.collection.ExternalAppendOnlyMapSuite$$anonfun$10$$anonfun$apply$mcV$sp$8.apply(ExternalAppendOnlyMapSuite.scala:216)
  at scala.collection.immutable.Stream.foreach(Stream.scala:594)
  ...
```

Do you think it's possible that somehow there's a race here in reporting the metrics? so that things are spilling but not reported to the listener by the time the number of spilled tasks is checked?

I could try putting in an `eventually` to let it check repeatedly for a short time? any better ideas, or reasons to think that's not it?

**Comment by andrewor14** [comment]
I see, that's possible. The right thing to do here is to add a `wait` of some sort in the listener to block until we have received the stage completed event. We've done this in some other test listeners as well. (I can't remember which ones off the top of my head).

**Pull request closed without merging**